subsys/net/lib/rest_client/src/rest_client.c

```diff
@@ -146,10 +145,12 @@ static int rest_client_sckt_timeouts_set(int fd)
 	return 0;
 }
 
-static int rest_client_sckt_connect(int *const fd, const char *const hostname,
-				     const uint16_t port_num,
-				     const sec_tag_t sec_tag,
-				     int tls_peer_verify)
+static int rest_client_sckt_connect(int *const fd,
+				    const char *const hostname,
+				    const uint16_t port_num,
+				    const sec_tag_t sec_tag,
+				    int tls_peer_verify,
+				    int32_t timeout_ms)
```

Comment:
If timeout occurs in http_client lib, I think that a null response is returned from callback (errno is not set). But if timeout occurs in sckt, I'd assume that recv() returns error, and in that case errno is set (no callback?). 
In both cases http_client_req() just return byte amount that was sent. But the difference is that, from sckt timeout we can get the errno and that might be good to return from rest_client in case of timeout. Although, we could (also) map a null response as a timeout in rest_client? 

Anyways, Can you have a trial how does it really works and what errno is in case of sckt timeout in rest_client http cb handler? I.e. would it be better to have shorter sckt timeout and http_client timeout as "perälauta"? Or do it both, map errno and null response as a timeout?

Reply:
As discussed, I've added check for zero HTTP status code into rest_client and that's mapped to `-ETIMEDOUT`. No errno mapping implemented as NULL response results into positive return value from `http_client_req` and we don't know if errno has been set by something else than http_client.

Reply:
I would not map NULL HTTP response as an error code. It's not technically a to-spec response from an HTTP server, but the REST JITP library relies on receiving a NULL response from nRF Cloud. NULL response is not the same as a timeout; With NULL response, we know that our HTTP request was received in full by the TCP endpoint. The TCP endpoint simply decided to close the connection (gracefully) without sending a response (not even an empty HTTP response payload). While this is abnormal for an HTTP transaction, it is not technically a connection error, and definitely distinct from a timeout

Reply:
A good point. So, JITP expects no response instead of ok/accepted. Yeah, that is abnormal. 
For me it looked like that null response from http_client lib represents: a timeout/no response, recv() fails / socket closed by remote peer.  I.e. I think that you cannot count on that request was really received by remote server without a http response.

But yeah, if changing this will break current rest jitp functionality and we cannot do this in this PR. Maybe eventually it would be better that http_client lib would be changed/addded to have an option by the app to distinguish all those different cases when waiting for a response. But that a topic of another discussion/PR.

Reply:
This discussion also occurs at comment https://github.com/nrfconnect/sdk-nrf/pull/7978#discussion_r905233649 so I'll close this one.